src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -2919,18 +2924,66 @@ public bool HasBoxingConversion(TypeSymbol source, TypeSymbol destination, ref H
             return false;
         }
 
-        internal static bool HasImplicitPointerConversion(TypeSymbol source, TypeSymbol destination)
+        internal static bool HasImplicitPointerToVoidConversion(TypeSymbol source, TypeSymbol destination)
         {
             Debug.Assert((object)source != null);
             Debug.Assert((object)destination != null);
 
             // SPEC: The set of implicit conversions is extended to include...
             // SPEC: ... from any pointer type to the type void*.
 
-            PointerTypeSymbol pd = destination as PointerTypeSymbol;
-            PointerTypeSymbol ps = source as PointerTypeSymbol;
-            return (object)pd != null && (object)ps != null && pd.PointedAtType.IsVoidType();
+            return source.IsPointerOrFunctionPointer() && destination is PointerTypeSymbol { PointedAtType: { SpecialType: SpecialType.System_Void } };
+        }
+
+#nullable enable
+        internal bool HasImplicitPointerConversion(TypeSymbol? source, TypeSymbol? destination, ref HashSet<DiagnosticInfo>? useSiteDiagnostics)
+        {
+            if (!(source is FunctionPointerTypeSymbol { Signature: { } sourceSig })
+                || !(destination is FunctionPointerTypeSymbol { Signature: { } destinationSig }))
+            {
+                return false;
+            }
+
+            if (sourceSig.ParameterCount != destinationSig.ParameterCount ||
+                sourceSig.CallingConvention != destinationSig.CallingConvention)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < sourceSig.ParameterCount; i++)
+            {
+                var sourceParam = sourceSig.Parameters[i];
+                var destintationParam = destinationSig.Parameters[i];
+
+                if (sourceParam.RefKind != destintationParam.RefKind)
+                {
+                    return false;
+                }
+
+                if (!hasVariantConversion(sourceParam.RefKind, destinationSig.Parameters[i].Type, sourceSig.Parameters[i].Type, ref useSiteDiagnostics))
+                {
+                    return false;
+                }
+            }
+
+            return sourceSig.RefKind == destinationSig.RefKind
+                   && hasVariantConversion(sourceSig.RefKind, sourceSig.ReturnType, destinationSig.ReturnType, ref useSiteDiagnostics);
+
+            bool hasVariantConversion(RefKind refKind, TypeSymbol sourceType, TypeSymbol destinationType, ref HashSet<DiagnosticInfo>? useSiteDiagnostics)
```

Comment:
>Variant [](start = 20, length = 7)

What is the "Variant" word supposed to convey?

Reply:
Co or Contravariant, depending on return or parameter context.

Reply:
> Co or Contravariant, depending on return or parameter context.

Consider dropping it from the name. 

---
In reply to: [402656947](https://github.com/dotnet/roslyn/pull/43007#discussion_r402656947) [](ancestors = 402656947)